Userland/Libraries/LibGUI/FileSystemModel.cpp

```diff
@@ -668,43 +671,66 @@ bool FileSystemModel::fetch_thumbnail_for(Node const& node)
         return true;
     }
 
-    // Otherwise, arrange to render the thumbnail
-    // in background and make it available later.
+    // See if the thumbnail was loaded, but the callback was never invoked.
+    // This often happens if a window with a FileSystemModel is closed and later reopened.
+    auto promise_it = s_loading_thumbnails.find(path);
+    if (promise_it != s_loading_thumbnails.end()) {
+        auto& result = promise_it->value->result();
+        // The promise was resolved since we started it.
+        if (result.has_value()) {
+            auto thumbnail_or_error = result.release_value();
+            s_loading_thumbnails.remove(path);
+
+            if (thumbnail_or_error.is_error()) {
+                s_thumbnail_cache.set(path, nullptr);
+                dbgln("Failed to load thumbnail for {}: {}", path, thumbnail_or_error.error());
+                return false;
+            }
+            s_thumbnail_cache.set(path, thumbnail_or_error.release_value());
+            node.thumbnail = s_thumbnail_cache.get(path).value();
+            return true;
+        }
+        return false;
+    }
+
+    // Otherwise, arrange to render the thumbnail in background and make it available later.
 
-    s_thumbnail_cache.set(path, nullptr);
     m_thumbnail_progress_total++;
 
     auto weak_this = make_weak_ptr();
 
-    (void)Threading::BackgroundAction<ErrorOr<NonnullRefPtr<Gfx::Bitmap>>>::construct(
-        [path](auto&) {
-            return render_thumbnail(path);
-        },
+    s_loading_thumbnails.set(path,
+        BitmapBackgroundAction::construct(
+            [path](auto&) {
+                return render_thumbnail(path);
+            },
 
-        [this, path, weak_this](auto thumbnail_or_error) -> ErrorOr<void> {
-            if (thumbnail_or_error.is_error()) {
-                s_thumbnail_cache.set(path, nullptr);
-                dbgln("Failed to load thumbnail for {}: {}", path, thumbnail_or_error.error());
-            } else {
-                s_thumbnail_cache.set(path, thumbnail_or_error.release_value());
-            }
+            [this, path, weak_this](auto& thumbnail_or_error) -> ErrorOr<void> {
```

Comment:
What's up with this API? Why would the action callback returning an Error ever call the completion function rather than the error function?

Reply:
Because the error callback only happens on cancel IIRC

Reply:
It seems to me that these callers passing `ErrorOr<T>` to their completion handler was a kludge because we didn't have an error callback.  "transparently" handling any error from the action and passing that to the error callback would make much more sense.